solver/src/settlement.rs

```diff
@@ -67,56 +62,70 @@ pub trait Interaction: std::fmt::Debug + Send {
     fn encode(&self) -> Vec<EncodedInteraction>;
 }
 
-#[derive(Debug, Default)]
-pub struct Settlement {
-    pub clearing_prices: HashMap<H160, U256>,
-    pub fee_factor: U256,
-    pub trades: Vec<Trade>,
-    pub pre_interactions: Vec<Box<dyn Interaction>>,
-    pub intra_interactions: Vec<Box<dyn Interaction>>,
-    pub post_interactions: Vec<Box<dyn Interaction>>,
+/// An intermediate settlement representation that can be incrementally
+/// constructed.
+///
+/// This allows liquidity to to encode itself into the settlement, in a way that
+/// is completely decoupled from solvers, or how the liquidity is modelled.
+/// Additionally, the fact that the settlement is kept in an intermediate
+/// representation allows the encoder to potentially perform gas optimizations
+/// (e.g. collapsing two interactions into one equivalent one).
+#[derive(Debug)]
+pub struct SettlementEncoder {
+    tokens: Vec<H160>,
+    clearing_prices: HashMap<H160, U256>,
+    trades: Vec<Trade>,
+    execution_plan: Vec<Box<dyn Interaction>>,
 }
 
-impl Settlement {
-    pub fn tokens(&self) -> Vec<H160> {
-        self.clearing_prices.keys().copied().collect()
+impl SettlementEncoder {
+    /// Creates a new settlement encoder with the specified prices.
+    ///
+    /// The prices must be provided up front in order to ensure that all tokens
+    /// included in the settlement are known when encoding trades.
+    fn new(clearing_prices: HashMap<H160, U256>) -> Self {
+        // Explicitely define a token ordering based on the supplied clearing
+        // prices. This is done since `HashMap::keys` returns an iterator in
+        // arbitrary order ([1]), meaning that we can't rely that the ordering
+        // will be consistent across calls. The list is sorted so that
```

Comment:
Alternatively we could use BTreeMap which is always sorted by key. But I see below we wouldn't be able to efficiently find the index (position) of a key with just the BTreeMap api.

Reply:
Since we create the vector once and sort it, I find it easier to just use a `Vec` since it has the same runtime characteristics. I think, that if we would insert into the list while building, then a binary tree might be nice.